tiledb/sm/query/reader_base.cc

```diff
@@ -414,228 +414,200 @@ Status ReaderBase::read_tiles(
   if (result_tiles->empty())
     return Status::Ok();
 
-  // Reading tiles are thread safe. However, we will perform
-  // them on this thread if there is only one read to perform.
-  if (names->size() == 1) {
-    RETURN_NOT_OK(read_tiles(names->at(0), result_tiles));
-  } else {
-    const auto status = parallel_for(
-        storage_manager_->compute_tp(),
-        0,
-        names->size(),
-        [&](const uint64_t i) {
-          RETURN_NOT_OK(read_tiles(names->at(i), result_tiles));
-          return Status::Ok();
-        });
-
-    RETURN_NOT_OK(status);
-  }
-
-  return Status::Ok();
-}
-
-Status ReaderBase::read_tiles(
-    const std::string& name,
-    const std::vector<ResultTile*>* result_tiles) const {
-  // Shortcut for empty tile vec
-  if (result_tiles->empty())
-    return Status::Ok();
-
-  // Read the tiles asynchronously
-  std::vector<ThreadPool::Task> tasks;
-  RETURN_CANCEL_OR_ERROR(read_tiles(name, result_tiles, &tasks));
-
-  // Wait for the reads to finish and check statuses.
-  auto statuses = storage_manager_->io_tp()->wait_all_status(tasks);
-  for (const auto& st : statuses)
-    RETURN_CANCEL_OR_ERROR(st);
-
-  return Status::Ok();
-}
-
-Status ReaderBase::read_tiles(
-    const std::string& name,
-    const std::vector<ResultTile*>* result_tiles,
-    std::vector<ThreadPool::Task>* const tasks) const {
-  // Shortcut for empty tile vec
-  if (result_tiles->empty())
-    return Status::Ok();
-
-  // For each tile, read from its fragment.
-  const bool var_size = array_schema_->var_size(name);
-  const bool nullable = array_schema_->is_nullable(name);
+  // Populate the list of regions per file to be read.
+  std::map<URI, std::vector<std::tuple<uint64_t, void*, uint64_t>>> all_regions;
+  std::mutex all_regions_mutex;
 
-  // Gather the unique fragments indexes for which there are tiles
-  std::unordered_set<uint32_t> fragment_idxs_set;
-  for (const auto& tile : *result_tiles)
-    fragment_idxs_set.emplace(tile->frag_idx());
+  // Mutex protecting attr_tiles_ maps in all ResultTiles.
+  std::mutex attr_tiles_mtx;
 
-  // Put fragment indexes in a vector
-  std::vector<uint32_t> fragment_idxs_vec;
-  fragment_idxs_vec.reserve(fragment_idxs_set.size());
-  for (const auto& idx : fragment_idxs_set)
-    fragment_idxs_vec.emplace_back(idx);
+  // Run all tiles and attributes in parallel.
+  auto status = parallel_for_2d(
+      storage_manager_->compute_tp(),
+      0,
+      names->size(),
+      0,
+      result_tiles->size(),
+      [&](uint64_t name_idx, unsigned rt_idx) {
+        auto& name = names->at(name_idx);
```

Comment:
`at` can throw if the index doesn't exist. I know here we know it will be fine but in general we should wrap in try/catch for exception handling.

Reply:
@Seth, would dereferencing the vector and using [] solve that issue?

Reply:
Well this is more about the goals of handling exceptions and making everything exception safe. Yes using `[]` avoid an exception since out_of_bounds behavior is undefined. However in this case it is effectively the same inside the 2D for.